packages/swapkit/api/src/swapkitApi/endpoints.ts

```diff
@@ -11,29 +12,99 @@ import {
   type TrackerParams,
   type TrackerResponse,
 } from "./types";
-
 const baseUrl = "https://api.swapkit.dev";
 const baseUrlDev = "https://dev-api.swapkit.dev";
 
 function getBaseUrl(isDev?: boolean) {
   return isDev ? baseUrlDev : baseUrl;
 }
 
-export function getTrackerDetails(payload: TrackerParams, apiKey?: string) {
-  return RequestClient.post<TrackerResponse>(`${getBaseUrl()}/track`, {
+export const computeHash = ({
+  method,
+  url,
+  apiKey,
+  payload,
+}: {
+  method: string;
+  url: string;
+  apiKey: string;
+  payload: any;
+}): string => {
+  switch (method) {
+    case "POST":
+      return computeHashForPost({ payload, apiKey });
+    case "GET":
+      return computeHashForGet({ url, apiKey });
+    default:
+      throw new SwapKitError("api_v2_invalid_method_key_hash", {
+        message: `Invalid method: ${method}`,
+      });
+  }
+};
+
+export const computeHashForGet = ({
+  url,
+  apiKey,
+}: {
+  url: string;
+  apiKey: string;
+}): string => {
+  return crypto.createHash("sha256").update(`${url}${apiKey}`, "utf8").digest("hex");
+};
+
+export const computeHashForPost = ({
+  apiKey,
+  payload,
+}: {
+  apiKey: string;
+  payload: any;
+}): string => {
+  const normalizedBody = JSON.stringify(payload);
+  return crypto.createHash("sha256").update(`${normalizedBody}${apiKey}`, "utf8").digest("hex");
+};
+
+export function getTrackerDetails(payload: TrackerParams, apiKey?: string, referer?: string) {
+  const url = `${getBaseUrl()}/track`;
+  const hash = apiKey
+    ? computeHash({
+        method: "POST",
+        apiKey: apiKey ?? "",
+        url,
+        payload,
+      })
+    : undefined;
+  return RequestClient.post<TrackerResponse>(url, {
     json: payload,
-    headers: apiKey ? { "x-api-key": apiKey } : {},
+    headers: {
+      ...(apiKey ? { "x-api-key": apiKey } : {}),
+      ...(hash ? { "x-payload-hash": hash } : {}),
+      ...(referer ? { referer } : {}),
+    },
   });
 }
 
 export async function getSwapQuoteV2<T extends boolean>(
   searchParams: QuoteRequest,
   isDev?: T,
   apiKey?: string,
+  referer?: string,
 ) {
-  const response = await RequestClient.post<QuoteResponse>(`${getBaseUrl(isDev)}/quote`, {
+  const url = `${getBaseUrl(isDev)}/quote`;
+  const hash = apiKey
```

Comment:
so when you have an api key you always compute the hash?
then why do we set the api key in the header?

Reply:
will update to exclude api key header if there is a hash